add support for sending cpu sparse tensors over rpc

gcramer23 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #62960
* __->__ #62794

This pr updates jit serialization to support pickling Sparse COO tensors. 
This pr updates message.cpp to support Sparse COO tensors.
A bug was filed a few years ago https://github.com/pytorch/pytorch/issues/30807.

I tested the fix by adding sparse tensor tests to rpc_test.py and dist_autograd_test.py.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @gcramer23 @gmagogsfm

Differential Revision: [D30608848](https://our.internmc.facebook.com/intern/diff/D30608848)

gcramer23 committed:
add support for sending cpu sparse tensors over rpc

[ghstack-poisoned]

gcramer23 committed:
Update on "add support for sending cpu sparse tensors over rpc"




[ghstack-poisoned]

gcramer23 committed:
Update on "add support for sending cpu sparse tensors over rpc"




[ghstack-poisoned]

gcramer23 committed:
Update on "add support for sending cpu sparse tensors over rpc"


This pr updates jit serialization to support sending CPU sparse tensors over RPC. A bug was filed a few years ago https://github.com/pytorch/pytorch/issues/30807.

I tested the fix by in rpc_test.py by adding a sparse tensor check in test_send_to_rank.

[ghstack-poisoned]

gcramer23 committed:
Update on "add support for sending cpu sparse tensors over rpc"


This pr updates jit serialization to support sending CPU sparse tensors over RPC. A bug was filed a few years ago https://github.com/pytorch/pytorch/issues/30807.

I tested the fix by in rpc_test.py by adding a sparse tensor check in test_send_to_rank.

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/message.cpp:
```diff
@@ -68,10 +68,25 @@ void Message::setId(int64_t id) {
-  storages.reserve(tensors_.size());
-  for (const auto& tensor : tensors_) {
-    storages.emplace_back(tensor.storage().getWeakStorageImpl());
+  int64_t storage_pointers = 0;
```
nit: var and method names in this file are using camel instead of snake convention.

gcramer23 replied:
Okay, that was a mistake.

mrshenli commented on torch/csrc/distributed/rpc/message.cpp:
```diff
@@ -68,10 +68,25 @@ void Message::setId(int64_t id) {
 
 std::vector<c10::weak_intrusive_ptr<c10::StorageImpl>> Message::getStorages()
     const {
+  // Sparse tensors do not have storage. Instead, a sparse tensor
```
curious about the design decision. Any reason we push the concept of SparseTensor to Message and JIT pickling, instead of extracting dense tensors from sparse tensors in RPC pickler:

https://github.com/pytorch/pytorch/blob/a01f83232946b13f3543f29d14848727ada15986/torch/distributed/rpc/internal.py#L52-L61

Is the concern mainly about performance? If yes, how much perf diff have we seen between the two options (edit RPC pickler vs edit JIT pickler)?

gcramer23 replied:
> curious about the design decision. Any reason we push the concept of SparseTensor to Message and JIT pickling, instead of extracting dense tensors from sparse tensors in RPC pickler:

The reason was to make a sparse tensor a picklable object. If the RPC pickler contains the logic for serialization and deserialization, then any other uses of the JIT pickler cannot pickle a sparse a tensor. Also, JIT pickler contains code for quantized tensors and it is not in the RPC pickler. Right now you can you send cpu quanitized tensors over RPC because they are picklable object. So the thought process was to make a sparse tensor picklable like a quantized tensor.

mrshenli replied:
I see, makes sense. I assume this does not affect TensorPipe at all, because although `rpc::Message` can now contain SparseTensors, TensorPipe only sees dense Tensors after `tensorpipeSerialization` in `tensorpipe_utis.cpp`?

gcramer23 replied:
Yes, TensorPipe still only sees dense Tensors.

mrshenli commented on torch/csrc/jit/serialization/pickler.h:
```diff
@@ -192,7 +193,6 @@ class TORCH_API Pickler {
       const std::string& class_name);
   // raw string data is appended directly to the byte stream
   void pushBytes(const std::string& string);
-  void pushTensorData(const at::Tensor& tensor);
```
Why deleting this line?

gcramer23 replied:
It is not used in the PyTorch repo / its deadcode

mrshenli replied:
Shall we move this change to a separate PR? We usually would avoid sneaking in irrelevant changes in PRs.

gcramer23 replied:
Okay, I'll make a pr for this.

mrshenli reviewed (commented):
Could you please rebase to clear the merge conflicts in the following tests?

https://app.circleci.com/pipelines/github/pytorch/pytorch/361709/workflows/6f73343e-6eec-4f98-8c9b-4ebfcf62d08d/jobs/15256178

gcramer23 commented:
> Could you please rebase to clear the merge conflicts in the following tests?
> 
> https://app.circleci.com/pipelines/github/pytorch/pytorch/361709/workflows/6f73343e-6eec-4f98-8c9b-4ebfcf62d08d/jobs/15256178

Okay Ill do that.

gcramer23 committed:
Update on "add support for sending cpu sparse tensors over rpc"


This pr updates jit serialization to support sending CPU sparse tensors over RPC. A bug was filed a few years ago https://github.com/pytorch/pytorch/issues/30807.

I tested the fix by in rpc_test.py by adding a sparse tensor check in test_send_to_rank.

[ghstack-poisoned]

mrshenli commented on torch/csrc/jit/serialization/pickler.cpp:
```diff
@@ -352,6 +352,40 @@ void Pickler::pushTensor(const IValue& ivalue) {
+  for (auto size : tensor.sizes()) {
+    pushInt(size);
+  }
+  push<PickleOpCode>(PickleOpCode::TUPLE);
```
curious, why do we need two `PickleOpCode::TUPLE`?

gcramer23 replied:
My understanding is that `PickleOpCode::TUPLE` is used after a `PickleOpCode::MARK` to create an immutable list.  In the `pushLiteralSparseTensor` since there are two `PickleOpCode::MARK` two `PickleOpCode::TUPLE` are required.

gcramer23 replied:
The first `PickleOpCode::MARK` is used for the sparse tensor. The second `PickleOpCode::MARK` is used for `tensor.sizes()`, the definition is `at::IntArrayRef sizes()`.

gcramer23 commented on torch/csrc/distributed/rpc/message.cpp:
```diff
@@ -68,10 +68,25 @@ void Message::setId(int64_t id) {
-  storages.reserve(tensors_.size());
-  for (const auto& tensor : tensors_) {
-    storages.emplace_back(tensor.storage().getWeakStorageImpl());
+  int64_t storagePtrs = 0;
```
Okay, I will fix it.

gcramer23 commented on torch/csrc/distributed/rpc/message.cpp:
```diff
@@ -68,10 +68,25 @@ void Message::setId(int64_t id) {
-  for (const auto& tensor : tensors_) {
-    storages.emplace_back(tensor.storage().getWeakStorageImpl());
+  int64_t storagePtrs = 0;
+  for (const at::Tensor& tensor : tensors_) {
+    if (tensor.is_sparse()) {
+      storagePtrs += 2;
+    } else {
+      storagePtrs += 1;
+    }
+  }
```
Okay.

gcramer23 commented on torch/csrc/jit/serialization/pickler.cpp:
```diff
@@ -352,6 +352,40 @@ void Pickler::pushTensor(const IValue& ivalue) {
 }
 
+void Pickler::pushLiteralSparseTensor(const at::Tensor& tensor) {
+  // The arguments to this function are:
```
I don't know if it's needed, but it potentially will make finding future problems easier. I'll add a TORCH_INTERNAL_ASSERT.

gcramer23 commented on torch/csrc/jit/serialization/unpickler.cpp:
```diff
@@ -647,6 +651,35 @@ void Unpickler::readGlobal(
+    // rebuild result tensor
+    at::Tensor result = at::_sparse_coo_tensor_unsafe(
+        indices_tensor, values_tensor, size, options);
+    result = autograd::make_variable(result, requires_grad);
```
`void Unpickler::rebuildTensor(bool quantized)` makes a variable 

```
  result = autograd::make_variable(result, requires_grad);
  stack_.emplace_back(std::move(result));
```

mrshenli replied:
I see, I missed that.

gcramer23 commented on torch/csrc/jit/serialization/unpickler.cpp:
```diff
@@ -647,6 +651,35 @@ void Unpickler::readGlobal(
+                       .dtype(data_type)
+                       .layout(c10::Layout::Sparse)
+                       .device(device)
+                       .pinned_memory(pinned_memory)
```
When rebuilding the dense tensor it uses the TensorOptions from the storage `      result = at::empty({0}, storage_tensor.options());
`.

Since sparse tensors do not have storage, I pushed the values required to build TensorOptions. TensorOptions is used when building the result sparse tensor.

gcramer23 replied:
> Is this same as sparse_tensor.values().options().layout(c10::Layout::Sparse)?
What do you mean?

This is for setting the options of the sparse tensor, not the values or indices.

mrshenli replied:
> What do you mean?

Yep, I am curious whether these two will return the same options value?

1. in comment
```c++
auto options = values_tensor.options().layout(c10::Layout::Sparse);
```

2.  in PR
```c++
bool requires_grad = elements.at(idx++).toBool();
bool pinned_memory = elements.at(idx++).toBool();
c10::ScalarType data_type =
        static_cast<c10::ScalarType>(elements.at(idx++).toInt());
auto options = c10::TensorOptions()
                       .dtype(data_type)
                       .layout(c10::Layout::Sparse)
                       .pinned_memory(pinned_memory)
                       .requires_grad(requires_grad);
```

gcramer23 committed:
Update on "add support for sending cpu sparse tensors over rpc"


This pr updates jit serialization to support sending CPU sparse tensors over RPC. A bug was filed a few years ago https://github.com/pytorch/pytorch/issues/30807.

I tested the fix by in rpc_test.py by adding a sparse tensor check in test_send_to_rank.

[ghstack-poisoned]

gcramer23 commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -641,10 +641,19 @@ def test_self_add(self):
             self.assertEqual(ret, torch.ones(2, 2) + 1)
 
+        # Test sparse tensor
+        for exec_mode in [RPCExecMode.SYNC, RPCExecMode.ASYNC, RPCExecMode.REMOTE]:
```
Okay, I added tests for autograd,  RPC, and parameter server application.

mrshenli commented on torch/testing/_internal/distributed/rpc/dist_autograd_test.py:
```diff
@@ -395,11 +421,9 @@ def _test_graph(self, fn, exec_mode):
                 ).to_here()
             else:
                 raise ValueError("Unrecognized ExecMode {}".format(exec_mode))
-
```
shall we revert unintentionally deleted lines?

gcramer23 replied:
I deleted these lines on purpose when going through the tests that I updated. I see no reason for having them. Do you want me to add them back?

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -4096,6 +4460,71 @@ def rref_error():
+
+    def _vanilla_ps(self, tensor):
+        # create rref on self
+        rref_self = rpc.remote(
+            worker_name(self.rank),
+            self.VanillaPs,
+            args=(self.world_size - 1,))
```
This can be replaced by the following?

```python
ps_rref = RRef(VanillaPs(self.world_size - 1))
```

gcramer23 replied:
Yes, I updated the code.

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -4096,6 +4460,71 @@ def rref_error():
 
         dist.barrier()
 
+    class VanillaPs:
```
For consistency with other test classes and method, let's just call it `MyParameterServer`

gcramer23 replied:
Okay.

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -4096,6 +4460,71 @@ def rref_error():
+        fut = rref.rpc_async().average(*[rref, tensor])
+        return fut.wait()
+
+    def _vanilla_ps(self, tensor):
```
this needs to change accordingly too.

gcramer23 replied:
Okay.

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -4096,6 +4460,71 @@ def rref_error():
+            return fut
+
+    def _trainer_func(self, rref, tensor):
+        fut = rref.rpc_async().average(*[rref, tensor])
```
any reason this is not just `fut = rref.rpc_async().average(rref, tensor)`?

gcramer23 replied:
I updated the code.

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -4096,6 +4460,71 @@ def rref_error():
+                        fut.set_result(result)
+            return fut
+
+    def _trainer_func(self, rref, tensor):
```
Shall we replace the trainer with a more realistic implementation, where there are sparse gradients from EmbeddingTables?

gcramer23 replied:
Okay, I will update the trainer.

mrshenli reviewed (commented):
Thanks @gcramer23 for adding extensive tests for sparse tensors. This LGTM. 

However, I do have one concern on the CI overhead. This PR will double the test time for the RPC module. @malfet is this OK? If this is a problem, we can try to pick a smaller representative subset of tests.

malfet commented on torch/csrc/distributed/rpc/message.cpp:
```diff
@@ -68,10 +68,23 @@ void Message::setId(int64_t id) {
-  for (const auto& tensor : tensors_) {
-    storages.emplace_back(tensor.storage().getWeakStorageImpl());
+  auto numStorages = tensors_.size();
+  for (const at::Tensor& tensor : tensors_) {
```
Nit
```suggestion
  for (const auto& tensor : tensors_) {
```

gcramer23 replied:
Okay.

malfet commented on torch/csrc/distributed/rpc/message.cpp:
```diff
@@ -68,10 +68,23 @@ void Message::setId(int64_t id) {
+    }
+  }
+  storages.reserve(numStorages);
+  for (const at::Tensor& tensor : tensors_) {
```
Nit
```suggestion
  for (const auto& tensor : tensors_) {
```

gcramer23 replied:
Okay.

malfet commented on torch/csrc/distributed/rpc/message.cpp:
```diff
@@ -68,10 +68,23 @@ void Message::setId(int64_t id) {
-  storages.reserve(tensors_.size());
-  for (const auto& tensor : tensors_) {
-    storages.emplace_back(tensor.storage().getWeakStorageImpl());
+  auto numStorages = tensors_.size();
+  for (const at::Tensor& tensor : tensors_) {
+    if (tensor.is_sparse()) {
+      ++numStorages;
+    }
+  }
+  storages.reserve(numStorages);
```
Just curious, does extra loop over tensors outweighs perf benefits of doing something like `storages.reserve(tensors_.size());` or `storages.reserver(2*tensors_.size());`?

gcramer23 replied:
I did a basic execution time experiment using chrono::high_resolution_clock. r2x measures the time it takes to reserve 2 * x elements. looprx measures the time it takes to loop x times and checks if a tensor is sparse. In the test, the tensor is not sparse.

```
1-r2x 227nanoseconds
1-looprx 96nanoseconds
10-r2x 135nanoseconds
10-looprx 73nanoseconds
100-r2x 2100nanoseconds
100-looprx 82nanoseconds
1000-r2x 2204nanoseconds
1000-looprx 1888nanoseconds
10000-r2x 4171nanoseconds
10000-looprx 3093nanoseconds
100000-r2x 4243nanoseconds
100000-looprx 5579nanoseconds
1000000-r2x 6117nanoseconds
1000000-looprx 4935nanoseconds
10000000-r2x 6836nanoseconds
10000000-looprx 4188nanoseconds
```

Worst case, the extra loop should be slower since it reserves the same amount. 

Below is measurements for a loop and check if a tensor is sparse.

```
1-loopx 21nanoseconds
10-loopx 21nanoseconds
100-loopx 21nanoseconds
1000-loopx 22nanoseconds
10000-loopx 22nanoseconds
100000-loopx 22nanoseconds
1000000-loopx 21nanoseconds
10000000-loopx 21nanoseconds
```

The loop doesn't cost much in terms of time. 

Also, I think dense tensors are used more than sparse tensors, so based on the basic experiments, I would not reserve 2x.

gcramer23 replied:
I did a basic execution time experiment using chrono::high_resolution_clock. r2x measures the time it takes to reserve 2 * x elements. looprx measures the time it takes to loop x times and checks if a tensor is sparse. In the test, the tensor is sparse:

```
1-loopx 28 nanoseconds
1-r2x 198 nanoseconds
10-loopx 21 nanoseconds
10-r2x 2514 nanoseconds
100-loopx 21 nanoseconds
100-r2x 77 nanoseconds
1000-loopx 21 nanoseconds
1000-r2x 1951 nanoseconds
10000-loopx 22 nanoseconds
10000-r2x 3536 nanoseconds
100000-loopx 22 nanoseconds
100000-r2x 3962 nanoseconds
1000000-loopx 21 nanoseconds
1000000-r2x 5841 nanoseconds
10000000-loopx 21 nanoseconds
10000000-r2x 5119 nanoseconds
```

Worst case, the extra loop should be slower since it reserves the same amount. 

The loop doesn't cost much in terms of time. 

Also, I think dense tensors are used more than sparse tensors, so based on the basic experiments, I would not reserve 2x.

gcramer23 replied:
I think there is an error in the loopx timing.

gcramer23 replied:
```
  if (count > 0){
        for (int i = 1; i < 10000000; i*=10){
          const auto& begin = std::chrono::high_resolution_clock::now();
          int ns = 1;
          for(int j = 0; j < i; j++){
            if (save.is_sparse()){
              ns++;
            }
          }
          const auto& end = std::chrono::high_resolution_clock::now();
          std::cout << i << "-loopx " << std::chrono::duration_cast<std::chrono::nanoseconds>(end - begin).count() << "nanoseconds" << std::endl;
        }
  }
 ```
 
 Idk I guess its right actually 
 ```
1-loopx 64nanoseconds
10-loopx 33nanoseconds
100-loopx 32nanoseconds
1000-loopx 33nanoseconds
10000-loopx 32nanoseconds
100000-loopx 32nanoseconds
1000000-loopx 32nanoseconds
1-loopx 41nanoseconds
10-loopx 33nanoseconds
100-loopx 33nanoseconds
1000-loopx 32nanoseconds
10000-loopx 33nanoseconds
100000-loopx 32nanoseconds
1000000-loopx 32nanoseconds
1-loopx 72nanoseconds
10-loopx 32nanoseconds
100-loopx 33nanoseconds
1000-loopx 32nanoseconds
10000-loopx 32nanoseconds
100000-loopx 33nanoseconds
1000000-loopx 32nanoseconds

```

malfet reviewed (commented):
@mrshenli what is the current wall time for testing RPC module?

mrshenli commented:
> @mrshenli what is the current wall time for testing RPC module?

@malfet for CPU test (this PR is CPU only), I check [this CI env](https://app.circleci.com/pipelines/github/pytorch/pytorch/364917/workflows/e3bf7724-1d5c-4895-91cb-4b02d7379b01/jobs/15396257) and it's about 15min. 

```
Ran 369 tests in 905.997s
```

gcramer23 committed:
Update on "add support for sending cpu sparse tensors over rpc"


This pr updates jit serialization to support sending CPU sparse tensors over RPC. A bug was filed a few years ago https://github.com/pytorch/pytorch/issues/30807.

I tested the fix by in rpc_test.py by adding a sparse tensor check in test_send_to_rank.

[ghstack-poisoned]

wanchaol commented on torch/csrc/jit/serialization/pickler.cpp:
```diff
@@ -353,6 +353,41 @@ void Pickler::pushTensor(const IValue& ivalue) {
+  // pinned memory
+  pushBool(tensor.options().pinned_memory());
+  // data type
+  pushInt(static_cast<uint16_t>(tensor.scalar_type()));
```
what about memory format? should we also save that or sparse_coo is the only format that sparse tensor gonna support?

gcramer23 replied:
The other sparse tensor format CSR. I did not update the serialization to support CSR. Right now the only sparse tensor format that will work is COO. So I did not save the memory format.

wanchaol reviewed (approved):
overall looks good to me on JIT related changes, some comments and suggestions inline.

wanchaol commented on torch/csrc/jit/serialization/pickler.cpp:
```diff
@@ -353,6 +353,41 @@ void Pickler::pushTensor(const IValue& ivalue) {
+  // data type
+  pushInt(static_cast<uint16_t>(tensor.scalar_type()));
+  // device
+  pushString(tensor.device().str());
```
the PR description mentioned we are adding CPU sparse tensor support, but here it looks like we are adding a generic pickle/unpickle support for tensor (i.e. in all devices), is this guarded somewhere in RPC side to ensure we are only sending cpu sparse tensor?

gcramer23 replied:
No, I did not add any guards. The next pr in the stack adds necessary changes for tensors on a CUDA device. 

A guard will cause the application to fail. So I don't understand why we would need a guard because the application will also fail without the guard.

gcramer23 commented on torch/csrc/distributed/rpc/message.cpp:
```diff
@@ -68,10 +68,23 @@ void Message::setId(int64_t id) {
+      ++numStorages;
+    }
+  }
+  storages.reserve(numStorages);
```
We use reserve instead of emplace_back to prevent vector reallocation. If the vector capacity is smaller than the new size, then automatic reallocation will occur.

```
1
2
4
8
16
32
64
128
```

The minimum size is N when all tensors are dense. The maximum size is 2N when all tensors are sparse.  Reserve runs in linear time https://en.cppreference.com/w/cpp/container/vector/reserve. 

Both are identical in Big O notation time complexity. I analyzed without counting constant operations such as emplace_back.

Solution 1 two loops:
The worst case for solution 1 is Loop N elements.  Reserve max 2N. Loop N tensors emplace_back.
O(4N) = O(N)
The best case for solution 1 is Loop N elements. Reserve N elements. Loop N tensors emplace_back.
O(3N) = O(N)

Solution 2 one loop reserve 2N:
The worst case is reserve 2N. Then loop N tensors emplace_back.
O(3N) = O(N)

Both are identical in Big O space complexity.  

I was leaning towards solution 1 because of the additional memory needed to run the program, not the algorithm. Also, I have the first use case for sparse tensors going through this code path, so I assume most tensors will be dense.

Rethinking this though I will merge the loops based on time complexity.

gcramer23 commented on torch/csrc/jit/serialization/pickler.cpp:
```diff
@@ -352,6 +352,40 @@ void Pickler::pushTensor(const IValue& ivalue) {
   }
 }
 
+void Pickler::pushLiteralSparseTensor(const at::Tensor& tensor) {
```
okay.

gcramer23 committed:
Update on "add support for sending cpu sparse tensors over rpc"


This pr updates jit serialization to support sending CPU sparse tensors over RPC. A bug was filed a few years ago https://github.com/pytorch/pytorch/issues/30807.

I tested the fix by in rpc_test.py by adding a sparse tensor check in test_send_to_rank.

[ghstack-poisoned]

mrshenli commented on torch/testing/_internal/distributed/rpc/dist_autograd_test.py:
```diff
@@ -64,13 +64,29 @@ def _torch_ones(sizes, requires_grad=False):
+    x = grads[rref.local_value()]
+    if x.is_sparse:
+        assert grad.is_sparse
+        x = x.to_dense()
+        grad = grad.to_dense()
```
Is this because `torch.equal` does not support sparse tensors and there is no `torch.sparse.equal`? If so, shall we create an issue to track?

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -4110,6 +4520,51 @@ def rref_error():
         dist.barrier()
 
+    def _trainer_func(self, rref, sparse):
+        m = MyEmbeddingBagModel(sparse=sparse)
+        loss_fn = nn.MSELoss()
+        outputs = m(torch.rand(10, 10).long())
+        loss_fn(outputs, torch.rand(10, 10)).backward()
+        gradient = list(m.parameters())[0].grad
+        fut = rref.rpc_async().average(rref, gradient)
+        return fut.wait()
```
Shall we run this for multiple iterations?

mrshenli commented on torch/testing/_internal/distributed/rpc/rpc_test.py:
```diff
@@ -4110,6 +4520,51 @@ def rref_error():
+        outputs = m(torch.rand(10, 10).long())
+        loss_fn(outputs, torch.rand(10, 10)).backward()
+        gradient = list(m.parameters())[0].grad
+        fut = rref.rpc_async().average(rref, gradient)
```
this can be done in a followup PR. Let's also test using RPC to sync gradients in a DDP comm hook, as that's the expected use case.

mrshenli reviewed (approved):
LGTM! Left some minor comments.

mrshenli commented:
@pytorchbot ciflow rerun -l ciflow/all

gcramer23 committed:
Update on "add support for sending cpu sparse tensors over rpc"


This pr updates jit serialization to support pickling Sparse COO tensors. 
This pr updates message.cpp to support Sparse COO tensors.
A bug was filed a few years ago https://github.com/pytorch/pytorch/issues/30807.

I tested the fix by adding sparse tensor tests to rpc_test.py and dist_autograd_test.py.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @gcramer23 @gmagogsfm

[ghstack-poisoned]

gcramer23 committed:
Update on "add support for sending cpu sparse tensors over rpc"


This pr updates jit serialization to support pickling Sparse COO tensors. 
This pr updates message.cpp to support Sparse COO tensors.
A bug was filed a few years ago https://github.com/pytorch/pytorch/issues/30807.

I tested the fix by adding sparse tensor tests to rpc_test.py and dist_autograd_test.py.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @gcramer23 @gmagogsfm

[ghstack-poisoned]

gcramer23 committed:
Update on "add support for sending cpu sparse tensors over rpc"


This pr updates jit serialization to support pickling Sparse COO tensors. 
This pr updates message.cpp to support Sparse COO tensors.
A bug was filed a few years ago https://github.com/pytorch/pytorch/issues/30807.

I tested the fix by adding sparse tensor tests to rpc_test.py and dist_autograd_test.py.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @gcramer23 @gmagogsfm

Differential Revision: [D30608848](https://our.internmc.facebook.com/intern/diff/D30608848)

[ghstack-poisoned]

gcramer23 commented:
@gcramer23 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30608848).
